backend/assignments/1/src/main/java/shrishti/example/Transaction.java

```diff
@@ -0,0 +1,109 @@
+package shrishti.example;
+import shrishti.example.Logger.logger;
+import java.util.Objects;
+import java.util.Random;
+
+public class Transaction {
+    public enum Type{
+        BUY,
+        SELL,
+        UPDATEPRICE,
+        ADDVOLUME
+    }
+    private  Parse.CoinData coin;
+    private Type onGoingTxn;
+    private String getBlockHash() {
+        String SALTCHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
+        StringBuilder transactionHash = new StringBuilder();
+        Random rnd = new Random();
+
+        for (double i = 0; i < 199999999; i++) {
+            i = i;
+        }
+        while (transactionHash.length() < 128) {
+            int index = (int) (rnd.nextFloat() * SALTCHARS.length());
+            transactionHash.append(SALTCHARS.charAt(index));
+        }
+        String hashCode = transactionHash.toString();
+        return "0x" + hashCode.toLowerCase();
+    }
+    public Transaction(Parse.CoinData coin, Type onGoingTxn){
+        this.coin=coin;
+        this.onGoingTxn=onGoingTxn;
+    }
+    public synchronized void buy() throws InterruptedException{
+        Trader trader = searchTrader(coin.getWalletAddress());
+        if(trader==null)return;
+        Coin newCoin = null;
+        for(Coin coinA : MarketPlace.getListOfCoins()){
+            if(Objects.equals(coinA.getSymbol(), coin.getCoin()))
+                newCoin=coinA;
+        }
+        while(newCoin!=null && !newCoin.buyCoins(trader, coin.getQuantity())){
```

Comment:
Add validations for all methods ina cleaner way. And abstract them out.

Wrap them in a validator wrapper/function.
What if these checks are general and reusable?